### PR TITLE
Compiler assertions according to issue #575

### PR DESCRIPTION
**Pull request by trappmartin**
This PR implements some simple assertions to allow for meaningful erroring of the compiler if the right-hand side of a tilde operant is not a Distribution or a Vector of Distributions.

### CODE REVIEW

File: src/core/compiler.jl [hunk 1]
```diff
@@ -35,6 +38,14 @@ Generate an observe expression.
 function generate_observe(observation, distribution)
     return esc(
         quote
+            # Sanity check.
+            isdist = if isa($(distribution), Vector)
```
Comment (willtebbutt) [inline src/core/compiler.jl]: It's not immediately obvious from the way that the `distribution` variable is named that it is allowed to be either a `Distribution` or `Vector` thereof. Could we please make this clearer in the docstring for each method that has this variable as an argument?

Comment (willtebbutt) [inline src/core/compiler.jl]: Also, `Vector` or `AbstractVector`?

File: src/core/compiler.jl [hunk 2]
```diff
@@ -83,6 +103,14 @@ function generate_assume(variable::Expr, distribution)
             indexing = mapfoldl(string, *, idcs, init = "")
             varname = Turing.VarName(vi, Symbol(csym_str), sym, indexing)
 
+            # Sanity check.
+            isdist = if isa($(distribution), Vector)
+                all(d -> isa(d, Distribution), $(distribution))
+            else
+                isa($(distribution), Distribution)
+            end
+            @assert isdist @error($(wrong_dist_errormsg))
```
Comment (willtebbutt) [inline src/core/compiler.jl]: This same block of code is repeated several times. Could we please put it in a function?

File: test/compiler.jl/model_macro.jl [hunk 3]
```diff
@@ -73,3 +73,17 @@ f01_mm = testmodel01()
...
+# Test for assertions.
+@model brokentestmodel(x1, x2) = begin
+    s ~ InverseGamma(2,3)
+    m ~ Normal(0,sqrt(s))
+
+    x1 ~ Normal(m, sqrt(s))
+    x2 ~ x1 + 2
+
+    return x1, x2
+end
```
Comment (willtebbutt) [inline test/compiler.jl/model_macro.jl]: This test is nice. Could we please implement one that will check both of the `generate_assume` methods? I believe that this one will only trigger the code in `generate_observe`.

File: src/core/compiler.jl (hunk 1 above)
Comment (trappmartin) [reply src/core/compiler.jl]: Good point!

File: test/compiler.jl/model_macro.jl (hunk 3 above)
Comment (trappmartin) [reply test/compiler.jl/model_macro.jl]: Correct, I'll add another test.

**Comment by yebai** [comment]
@trappmartin is this ready for merging now?

**Comment by trappmartin** [comment]
I would say yes but there seems to be something breaking on the Documenter part. Not sure where this comes from.

cc: @cpfiffer

**Comment by cpfiffer** [comment]
I think I know what's going on here, I'll fix it when I'm back at a computer.

**Comment by cpfiffer** [comment]
All good. Should be ready to merge this over now.

**Comment by trappmartin** [comment]
Thanks!

**Pull request merged**